### ISSUE TITLE
Accessibility issue using multiple drop-down links on menu

### ISSUE THREAD

**Issue by bobbulman** [localgovdrupal/localgov_microsites_base#219]
The built-in accessibility checker is reporting an issue with duplicate IDs when a menu has more than one drop-down column:

![image](https://github.com/localgovdrupal/localgov_microsites_base/assets/67878800/ac1721ed-14c5-4d20-8c7a-985eebcc1a33)

**Comment by markconroy** [comment]
@msayoung I think we'll need to do a bit of work on this.

If you have more than one sub menu at the same level, then we have duplicate ids.

For example:
```
Top Level 1 ................ Top Level 2 ................ Top Level 3
-- Submenu 1               -- Submenu 2              -- Submenu 3

```
The button to toggle each of those sub menu items will have the same id, since it takes its id from the `random(1000)` that is the id of the main `ul` for the menu.

Similarly
```
Top Level
-- Submenu 1
------ Subsubmenu 1
-- Submenu 2
------ Subsubmenu 2
-- Submenu 3
------ Subsubmenu 3
```
The Submenu 1/2/3 will all have the same ids.

However, worse than that, we don't have any `aria-controls` set on the button for each toggle (which was the reason we were using the random ids in the first place), so there will be an announcement that the button is "aria-expanded" but won't say what it was expanding.

We have a really nice "disclosure library" written in Annertech that perhaps we can use for this. You can see it doing lots of things here: https://annertech.pages.anner.ie/libraries/disclosure/

Using that library would go a long way towards fixing this, but could also be used for our accordions and anything else that opens/closes.

**Comment by msayoung** [comment]
Ooooh... lovely :) 

Thanks @bobbulman for spotting and reporting.

Would the reported issue be solved by making sure we are getting unique ids for each submenu? I'll fire up my local and have a test. 

The disclosure library looks great from quick glance. 
What are the potential ramifications of changing libraries at this stage, and would we be looking to do so in localgov proper too?

**Comment by markconroy** [comment]
> Would the reported issue be solved by making sure we are getting unique ids for each submenu? I'll fire up my local and have a test.

Yes, I believe so. But I've been having issues trying to get a way to give them a unique id AND have the parent trigger/button to have the same id as a data attribute or something. Let's have a call about this as soon as we can and see can we figure it out. It's a pretty simple issue once we can do that I hope.

**Comment by finnlewis** [comment]
@RyanFH257 just flagged this again, on the latest site. https://www.hfsafeguardingadults.org.uk/

**Comment by RyanFH257** [comment]
![image](https://github.com/user-attachments/assets/67735680-0a40-4d61-a44c-ea94a6a22beb)

**Comment by msayoung** [comment]
FYI - This isn't being flagged as an issue in automated tests anymore, probably because of the removal of https://www.w3.org/WAI/WCAG21/Understanding/parsing.html from WCAG 2.2

Duplicate IDs is still invalid HTML and could potentially create problems so I think its a good idea to fix.

**Comment by finnlewis** [comment]
Might this be useful? 
https://git.drupalcode.org/project/adaptivetheme/-/commit/c70534d89d2c360248634f0b1b4ff500706daece

**Comment by markconroy** [comment]
I can have a look at this again and see if Finn's link above is any help to us.

I have a feeling it won't be, because we need to have the same identifier on a parent as well (iirc).

**Comment by markconroy** [comment]
I think our recent updates to the menu/sub-menu (hover vs focus) may have solved this? 

I'm looking at a local site and I have 3 levels of sub menus, but everything seems to have unique ids.

```
<ul class="menu menu--group_menu_link_content-2 menu--menu-level-0" id="menu--237">
                    <li class="menu-item menu-item--level-0 menu-item--expanded" data-index="1" data-level="0">
        <a href="/" data-drupal-link-system-path="<front>">Home</a>
                  <button id="menu__sub-menu-toggle--237" data-id="237" class="menu__sub-menu-toggle" aria-label="Toggles Home submenu" aria-expanded="false">
            <span class="menu__sub-menu-toggle-icon"></span>
          </button>
                    
        
    
    
            <ul class="menu menu--sub-menu menu--sub-menu-level-1 menu---sub-menu" id="menu--611">
                    <li class="menu-item menu-item--level-1 menu-item--expanded" data-index="1" data-level="1">
        <a href="/about">About Us</a>
                  <button id="menu__sub-menu-toggle--611" data-id="611" class="menu__sub-menu-toggle" aria-label="Toggles About Us submenu" aria-expanded="false">
            <span class="menu__sub-menu-toggle-icon"></span>
          </button>
                    
        
    
    
            <ul class="menu menu--sub-menu menu--sub-menu-level-2 menu---sub-menu" id="menu--191">
                    <li class="menu-item menu-item--level-2 menu-item--expanded" data-index="1" data-level="2">
        <a href="/node" data-drupal-link-system-path="node">Sub 1</a>
                  <button id="menu__sub-menu-toggle--191" data-id="191" class="menu__sub-menu-toggle" aria-label="Toggles Sub 1 submenu" aria-expanded="false">
            <span class="menu__sub-menu-toggle-icon"></span>
          </button>
                    
        
    
    
            <ul class="menu menu--sub-menu menu--sub-menu-level-3 menu---sub-menu" id="menu--130">
                    <li class="menu-item menu-item--level-3" data-index="1" data-level="3">
        <a href="/node" data-drupal-link-system-path="node">Sub 2</a>
              </li>
        </ul>
  
              </li>
        </ul>
  
              </li>
        </ul>
  
              </li>

        </ul>
```

![Screenshot 2024-11-29 at 10 12 46](https://github.com/user-attachments/assets/c3a8fe7c-e950-48d9-853d-4e5562524bca)

If this is the case, can we close this issue? @bobbulman are you still getting this issue on your site(s)?

**Comment by msayoung** [comment]
Took another look at this on a new install, and on the H&F site above. 

If  we have two level 1 items, both with child items, then both top level items have the same ID. 
Equally, if an item has two children both with subitems, the toggle button also have the same ID.  see the screenshot below. 

<img width="1154" height="362" alt="Image" src="https://github.com/user-attachments/assets/e80823bd-93d5-4299-8726-101e8b4583ce" />

I'm not sure if we actually need the ID on the button at all. Can you think of a reason to keep it @markconroy ?

It is possible people will use it for styling (bad practice, but not unheard of) so whether we remove it, or change it we may need to flag this to people. 


I've just gone down a long rabbit hole about aria-controls. While generally it is common that if the expanded content is immediately after the control then it is acceptable to drop aria-controls. However it is still better to have it. It has scant support in screenreaders, but if it worked it would be useful . (In Jaws you can press a key combo to move to the controlled content). 

We do have unique ids on the controlled element (the menu <ul> ) , which we can use for the aria-controls. 


TLDR: the remaining issues here are 
- Duplicate IDs on sibling buttons 
- there is no aria-controls on the buttons

**Comment by msayoung** [comment]
Moved: https://git.drupalcode.org/project/localgov_microsites_base/-/work_items/3591369
